prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+	if (!qp) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Reservation QP create failed", ret);
+		return NULL;
+	}
+	return qp;
+}
+
+void fi_ibv_release_qpn(struct ibv_qp *rsvd_qp)
+{
+	ibv_destroy_qp(rsvd_qp);
+}
+
+static int fi_ibv_create_ini_qp(struct fi_ibv_ep *ep)
+{
+	struct ibv_qp_init_attr_ex attr_ex;
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	int ret;
+
+	fi_ibv_msg_ep_get_qp_attr(ep, NULL,
+			(struct ibv_qp_init_attr *)&attr_ex);
+	attr_ex.qp_type = IBV_QPT_XRC_SEND;
+	attr_ex.comp_mask = IBV_QP_INIT_ATTR_PD;
+	attr_ex.pd = domain->pd;
+	attr_ex.qp_context = domain;
+
+	ret = rdma_create_qp_ex(ep->id, &attr_ex);
+	if (ret) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "XRC INI QP, rdma_create_qp_ex()", -ret);
+		return ret;
+	}
+	return FI_SUCCESS;
+}
+
+static inline void fi_ibv_set_ini_conn_key(struct fi_ibv_ep *ep,
+					   struct fi_ibv_ini_conn_key *key)
+{
+	key->addr = ep->info->dest_addr;
+	key->tx_cq = container_of(ep->util_ep.tx_cq, struct fi_ibv_cq, util_cq);
+}
+
+/* Caller must hold domain:xrc:ini_mgmt_lock */
+struct fi_ibv_ini_shared_conn *fi_ibv_get_shared_ini_conn(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	struct fi_ibv_ini_conn_key key;
+	struct fi_ibv_ini_shared_conn *ini_conn;
+	struct ofi_rbnode *node;
+	int ret;
+
+	assert(ep->id);
+
+	fi_ibv_set_ini_conn_key(ep, &key);
+	node = ofi_rbmap_find(domain->xrc.ini_conn_rbmap, &key);
+	if (node) {
+		ini_conn = node->data;
+		ofi_atomic_inc32(&ini_conn->ref_cnt);
+		return ini_conn;
+	}
+
+	ini_conn = calloc(1, sizeof(*ini_conn));
+	if (!ini_conn) {
+		errno = FI_ENOMEM;
+		return NULL;
+	}
+
+	ini_conn->tgt_qpn = FI_IBV_NO_INI_TGT_QPNUM;
+	ini_conn->peer_addr = mem_dup(key.addr, ofi_sizeofaddr(key.addr));
+	if (!ini_conn->peer_addr) {
+		free(ini_conn);
+		errno = FI_ENOMEM;
+		return NULL;
+	}
+	ini_conn->tx_cq = container_of(ep->util_ep.tx_cq,
+				       struct fi_ibv_cq, util_cq);
+	dlist_init(&ini_conn->pending_list);
+	dlist_init(&ini_conn->active_list);
+	ofi_atomic_initialize32(&ini_conn->ref_cnt, 1);
+
+	ret = ofi_rbmap_insert(domain->xrc.ini_conn_rbmap,
+			       (void *) &key, (void *) ini_conn);
+	assert(ret != -FI_EALREADY);
+	if (ret) {
+		VERBS_WARN(FI_LOG_FABRIC, "INI QP RBTree insert failed %d\n",
+			   ret);
+		errno = -ret;
+		goto insert_err;
+	}
+	return ini_conn;
+
+insert_err:
+	free(ini_conn->peer_addr);
+	free(ini_conn);
+	return NULL;
+}
+
+/* Caller must hold domain:xrc:ini_mgmt_lock */
+void fi_ibv_put_shared_ini_conn(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	struct fi_ibv_ini_shared_conn *ini_conn = ep->ini_conn;
+	struct fi_ibv_ini_conn_key key;
+	struct ofi_rbnode *node;
+
+	if (!ini_conn)
+		return;
+
+	/* remove from pending or active connection list */
+	dlist_remove(&ep->ini_conn_entry);
+	ep->conn_state = FI_IBV_XRC_UNCONNECTED;
+	ep->ini_conn = NULL;
+	ep->ibv_qp = NULL;
```

Comment:
This ends up looking like an error.  Where did ibv_qp go if it wasn't already null?  Did we lose a reference to the qp here?

Reply:
The fi_ibv_ini_shared_conn object owns the XRC INI QP, not the endpoint that originally creates it. The fi_ibv_ini_shared_conn is reference counted and will destroy the physical XRC INI QP when all references are removed. See the lines right below these.